futures-util/src/option/mod.rs

```diff
@@ -0,0 +1,193 @@
+//! assert!(f.is_some());
+//! assert_eq!(f.current().await, 1);
+//! assert!(f.is_none());
+//! assert!(f.current().is_terminated());
+//! # });
+//! ```
+//!
+//! This is useful when you want to implement optional branches using the
+//! `select!` macro.
+//!
+//! ```
+//! #![feature(async_await)]
+//! #![recursion_limit="128"]
+//!
+//! use futures::{future, stream, OptionExt as _, StreamExt as _};
+//! # futures::executor::block_on(async {
+//! let mut value = None;
+//! let mut values = Some(stream::iter(vec![1u32, 2u32, 4u32].into_iter()).fuse());
+//! let mut parked = None;
+//!
+//! let mut sum = 0;
+//!
+//! loop {
+//!     futures::select! {
+//!         value = value.current() => {
+//!             sum += value;
+//!             std::mem::swap(&mut parked, &mut values);
+//!         }
+//!         v = values.next() => {
+//!             match v {
+//!                 Some(v) => {
+//!                     value = Some(async move { v });
+//!                     std::mem::swap(&mut parked, &mut values);
+//!                 },
+//!                 None => break,
+//!             }
+//!         }
+//!     }
+//! }
+//!
+//! assert_eq!(7, sum);
+//! # });
+//! ```
+//!
+//! [`Option`]: Option
+//! [`Stream`]: futures_core::stream::Stream
+//! [`Future`]: futures_core::future::Future
+
+use core::pin::Pin;
+use futures_core::{
+    future::{FusedFuture, Future},
+    stream::Stream,
+    task::{Context, Poll},
+};
+
+impl<T> OptionExt<T> for Option<T> {
+    fn next(&mut self) -> Next<'_, T>
+    where
+        T: Stream,
+    {
+        Next { stream: self }
+    }
+
+    fn current(&mut self) -> Current<'_, T>
+    where
+        T: Future,
+    {
+        Current { future: self }
+    }
+}
+
+/// Extension methods for [`Option`] of [`Stream`]s or [`Future`]s.
+///
+/// [`Option`]: std::option::Option
+/// [`Stream`]: Stream
+/// [`Future`]: Future
+pub trait OptionExt<T>: Sized {
+    /// Convert [`Option`] into a Future that resolves to the next item in the [`Stream`].
+    ///
+    /// If the [`Option`] is [`None`], the returned future also resolves to [`None`].
+    ///
+    /// [`Option`]: Option
+    /// [`None`]: Option::None
+    /// [`Stream`]: Stream
+    fn next(&mut self) -> Next<'_, T>
+    where
+        T: Stream;
+
+    /// Convert [`Option`] into a Future that resolves to the same value as the stored [`Future`].
+    ///
+    /// If the [`Option`] is [`None`], the returned future also resolves to [`None`].
+    ///
+    /// [`Option`]: Option
+    /// [`None`]: Option::None
+    /// [`Future`]: Future
+    fn current(&mut self) -> Current<'_, T>
+    where
+        T: Future;
+}
+
+/// Adapter future for `Option` to get the next value of the stored future.
+///
+/// Resolves to `None` if no future is present.
+#[derive(Debug)]
+pub struct Next<'a, T> {
+    pub(crate) stream: &'a mut Option<T>,
+}
+
+impl<'a, T> Next<'a, T> {
+    fn stream<'b>(self: Pin<&'b mut Self>) -> Pin<&'b mut Option<T>> {
```

Comment:
This seems to be getting `Pin<Field>` from `Pin<&mut Struct>`. I don't think it is safe to do this without the `Field: Unpin` bounds.

Reply:
Note that `Field` in this case is an `&mut Option<T> where T: Future`. The fact that we have hold mutable reference to the field should prevent it from moving for the duration of the projected `Pin`, no?

I'm very much unsure about this though.

Reply:
Indeed, this projection isn't OK. I've rebased and reverted the `Unpin` change. Thanks!